crates/circuit/src/dag_circuit.rs

```diff
@@ -6671,6 +6671,25 @@ impl DAGCircuit {
     ) -> PyResult<DAGCircuit> {
         // Extract necessary attributes
         let qc_data = qc.data;
+        Self::from_circuit_data(
+            &qc_data,
+            copy_op,
+            qc.name,
+            qc.metadata.map(|x| x.unbind()),
+            qubit_order,
+            clbit_order,
+        )
+    }
+
+    /// Builds a [DAGCircuit] based on an instance of [CircuitData].
+    pub fn from_circuit_data(
+        qc_data: &CircuitData,
+        copy_op: bool,
+        name: Option<String>,
+        metadata: Option<PyObject>,
+        qubit_order: Option<Vec<Bound<PyAny>>>,
+        clbit_order: Option<Vec<Bound<PyAny>>>,
+    ) -> PyResult<Self> {
```

Comment:
Why not use the native `Qubit`/`Clbit` type for the qubit/bit ordering here?

Reply:
I was just reusing what `from_circuit` did. But I think your right we can change this to be `ShareableQubit` and `ShareableClbit` as that's what these get normalized to. Since those have `FromPyObject` it should all just work fine.

Reply:
Fixed in: https://github.com/Qiskit/qiskit/pull/14668/commits/318634b2afac6c57d6494ee7cc8c27a8db02c1ba